lix/packages/client/src/git-http/optimizeReq.ts

```diff
@@ -0,0 +1,189 @@
+import {
+	decodeGitPackLines,
+	addBlobNoneFilter,
+	overrideHaves,
+	addWantsCapabilities,
+	overrideWants,
+	encodePackLine,
+	addNoProgress,
+} from "./helpers.js"
+
+/***
+ * This takes the request, decodes the request body and extracts each line in the format of the git-upload-pack protocol (https://git-scm.com/docs/gitprotocol-v2)
+ * and allows us to rewrite the request to add filters like blob:none or request only specific oids (overrideWants) or block list specific oids (overrideHaves)
+ */
+export async function optimizeReq(
+	gitConfig: {
+		noBlobs?: boolean
+
+		addRefs?: string[]
+
+		overrideHaves?: string[]
+		overrideWants?: string[]
+	} = {},
+	{
+		method,
+		url,
+		body,
+	}: {
+		method: string
+		url: string
+		body?: any
+	}
+) {
+	// Optimize refs requests, GET info/refs?service=git-upload-pack is a request to get the refs but does not support filtering, we rewrite this to a post upload pack v2 request that allows filetering refs how we want
+	// FIXME: document url + method api, how to get tree for commit just filter file blobs
+	// "http://localhost:3001/git-proxy//github.com/inlang/example/info/refs?service=git-upload-pack"
+	if (url.endsWith("info/refs?service=git-upload-pack") && gitConfig.addRefs !== undefined) {
+		const uploadPackUrl = url.replace("info/refs?service=git-upload-pack", "git-upload-pack")
+		const lines = []
+
+		// TODO: #1459 check if we have to ask for the symrefs
+		lines.push(encodePackLine("command=ls-refs"))
+
+		// 0001 - Delimiter Packet (delim-pkt) - separates sections of a message
+		lines.push(encodePackLine("agent=lix") + "0001")
+
+		if (gitConfig.addRefs?.length > 0) {
+			for (let ref of gitConfig.addRefs) {
+				if (!ref.startsWith("refs/")) {
+					ref = "refs/heads/" + ref
+				}
+				lines.push(encodePackLine("ref-prefix " + ref))
+			}
+		}
+
+		lines.push(encodePackLine("ref-prefix HEAD"))
+
+		// TODO: what does this do?
```

Comment:
This will add the HEAD symref to the response. compare https://git-scm.com/docs/protocol-v2#_ls_refs we need this since we replace the `info/refs?service=git-upload-pack` get request that doesn't support filtering with `git-upload-pack` post request. The response of the get request expects HEAD beeing part of the response - compare https://github.com/opral/monorepo/blob/447218286378e169394e3ef08428eb41674af3e4/lix/packages/client/src/git-http/optimizeReq.ts#L157. so we always add `symrefs` and the allowed prefix head via`ref-prefix head` to the request